std/digest/murmurhash.d

```diff
@@ -0,0 +1,789 @@
+}
+
+///
+unittest
+{
+    // One can also hash ubyte data piecewise.
+    const(ubyte)[] data1 = [1, 2, 3];
+    const(ubyte)[] data2 = [4, 5, 6, 7];
+    MurmurHash3_x86_32 hasher;
+    hasher.put(data1);
+    hasher.put(data2);
+    auto hashed = hasher.finish();
+}
+
+///
+unittest
+{
+    // Using SMurmurHash3_x86_32, SMurmurHash3_x86_128 and SMurmurHash3_x64_128
+    // you gain full control over which part of the algorithm to run.
+    // This allows for maximum throughput but needs extra care.
+
+    // Data type must be the same as the hasher's element type:
+    // - uint for SMurmurHash3_x86_32
+    // - ulong[2] for SMurmurHash3_x86_128 and SMurmurHash3_x64_128
+    const(uint)[] data = [1, 2, 3, 4];
+    // Note the hasher starts with S.
+    SMurmurHash3_x86_32 hasher;
+    // Push as many array of elements as you need. The less call the better performance wise.
+    hasher.putBlocks(data);
+    // Put remainder bytes if needed. This method can be called only once.
+    hasher.putRemainder(ubyte(1), ubyte(1), ubyte(1));
+    // Call finalize to incorporate data length in the hash.
+    hasher.finalize();
+    // Finally get the hashed value.
+    auto hashed = hasher.getBytes();
+}
+
+/// Implements MurmurHash3_x86_32 $(D std.digest.digest) Template API.
+alias MurmurHash3_x86_32 = Piecewise!SMurmurHash3_x86_32;
+/// Implements MurmurHash3_x86_128 $(D std.digest.digest) Template API.
+alias MurmurHash3_x86_128 = Piecewise!SMurmurHash3_x86_128;
+/// Implements MurmurHash3_x64_128 $(D std.digest.digest) Template API.
+alias MurmurHash3_x64_128 = Piecewise!SMurmurHash3_x64_128;
+
+/// Implements MurmurHash3_x86_32 $(D std.digest.digest.Digest) OOO API.
+alias MurmurHash3_x86_32Digest = WrapperDigest!MurmurHash3_x86_32;
+/// Implements MurmurHash3_x86_128 $(D std.digest.digest.Digest) OOO API.
+alias MurmurHash3_x86_128Digest = WrapperDigest!MurmurHash3_x86_128;
+/// Implements MurmurHash3_x64_128 $(D std.digest.digest.Digest) OOO API.
+alias MurmurHash3_x64_128Digest = WrapperDigest!MurmurHash3_x64_128;
+
+// This definition of NO_UNALIGNED_ACCESS is too restrictive. Only a few old
+// SPARC/ARM chips cannot do unaligned reads.
+version(ARM)   { version = NO_UNALIGNED_ACCESS; }
+version(SPARC) { version = NO_UNALIGNED_ACCESS; }
+
+/**
+Pushes an array of blocks at once. It is more efficient to push as much data as possible in a single call.
+On platform that does not support unaligned reads (some old ARM chips), it is forbidden to pass non aligned data.
+
+Note: Implementation uses pointer manipulation instead of foreach to avoid
+bounds checking (see @@@BUG@@@ 15581) making the function @trusted instead of
+@safe.
+This function heavily affects the performance of hash calculation so make sure
+to benchmark all changes. Benchmark can be found on
+$(WEB https://github.com/gchatelet/murmurhash3d, github).
+*/
+void putBlocks(H, Block = H.Block)(ref H hasher, scope const(Block[]) blocks...) pure nothrow @nogc @trusted
```

Comment:
This is misuse of `@trusted` - don't put it on templated functions which memory safety depends on template arguments. This silently breaks `@safe`'s promise of memory safety when `H.putBlock` can incur memory errors. I've written more about this [here](https://jakobovrum.github.io/d/2016/01/20/memory-safety.html).

Is `H` assumed to be a struct from this module? If so, and if all their `putBlock` functions have been manually vetted for memory safety when called in this manner, then I think it's fine as long as a template constraint is added to enforce the assumption. That way, all possible instantiations of `putBlocks` are memory safe, so it can correctly be annotated with `@trusted`.

However, if the `putBlock` functions all have a memory safe API, the approach using nested `@trusted` functions described in the aforementioned article is probably better, as it leaves `@trusted` to each individual `putBlock` implementation, making memory safety easier to review. Of course we'd have to see how it works with the benchmarks.

Also, if you don't want to deal with using `@trusted` correctly right now, feel free to leave things `@system`. It's better to leave memory safety guarantees to future work than violating `@safe`.